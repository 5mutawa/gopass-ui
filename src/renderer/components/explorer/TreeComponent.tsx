import * as React from 'react'
import * as t from 'react-treebeard'
import * as m from 'react-materialize'
import { customStyle, globalStyle } from './TreebeardStyle'

export interface Child {
    name: string
    children?: Child[]
    loading?: boolean
    entryId?: any
}

export interface Tree {
    name: string
    toggled: boolean
    children?: Child[]
}

export interface TreeComponentProps {
    tree: Tree
    onLeafClick: (leafId: string) => void
}

const decorators = {
    ...t.decorators
}

decorators.Header = ({ style, node }: any) => {
    const iconType = node.children ? 'folder' : 'insert_drive_file'
    const iconStyle = { marginRight: '5px' }

    return (
        <div style={style.base}>
            <div style={style.title}>
                <m.Icon small style={iconStyle}>
                    {iconType}
                </m.Icon>

                {node.name}
            </div>
        </div>
    )
}

export default class TreeComponent extends React.Component<TreeComponentProps, any> {
    constructor(props: TreeComponentProps) {
        super(props)
        this.state = {}
        this.onToggle = this.onToggle.bind(this)
    }

    onToggle(node: any, toggled: boolean) {
        if ((!node.children || node.children.length === 0) && !!node.entryId) {
            this.props.onLeafClick(node.entryId)
        }

        if (this.state.cursor) {
            this.state.cursor.active = false
        }
        node.active = true

        if (node.children) {
            node.toggled = toggled
        }
        this.setState({ cursor: node })
    }

    render() {
        return (
            <div>
                <t.Treebeard
                    data={this.props.tree}
                    decorators={decorators}
                    onToggle={this.onToggle}
                    style={globalStyle}
                />
            </div>
        )
    }
}
