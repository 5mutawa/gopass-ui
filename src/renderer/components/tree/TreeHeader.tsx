import * as React from 'react'
import * as m from 'react-materialize'

import { deriveIconFromSecretName } from '../../secrets/deriveIconFromSecretName'

export const TreeHeader = ({ style, node }: any) => {
    let iconType = node.toggled ? 'chevron_right' : 'folder'

    if (!node.children && node.path) {
        iconType = deriveIconFromSecretName(node.name)
    }

    return (
        <div style={style.base}>
            <m.Icon small>{iconType}</m.Icon>

            {node.name}
        </div>
    )
}
