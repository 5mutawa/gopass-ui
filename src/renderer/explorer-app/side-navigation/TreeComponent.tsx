import * as React from 'react'
import * as t from 'react-treebeard'
import * as m from 'react-materialize'
import { globalStyle } from './TreebeardStyle'
import credentialIconMappings, { CredentialIconMapping } from './CredentialIconMappings'

export interface Tree {
    name: string
    toggled?: boolean
    loading?: boolean
    children?: Tree[]
    entryId?: any
}

export interface TreeComponentProps {
    tree: Tree
    onLeafClick: (leafId: string) => void
}

t.decorators.Header = ({ style, node }: any) => {
    let iconType = node.toggled ? 'chevron_right' : 'folder'

    if (!node.children && node.entryId) {
        iconType = 'comment'

        credentialIconMappings.forEach((mapping: CredentialIconMapping) => {
            const hasIconMapping = mapping.regex.test(node.name)

            if (hasIconMapping) {
                iconType = mapping.icon
            }
        })
    }

    return (
        <div style={ style.base }>
            <m.Icon small>
                { iconType }
            </m.Icon>

            { node.name }
        </div>
    )
}

export default class TreeComponent extends React.Component<TreeComponentProps, any> {
    constructor(props: TreeComponentProps) {
        super(props)
        this.state = {}
        this.onToggle = this.onToggle.bind(this)
    }

    onToggle(node: any, toggled: boolean) {
        if ((!node.children || node.children.length === 0) && !!node.entryId) {
            this.props.onLeafClick(node.entryId)
        }

        if (this.state.cursor) {
            this.state.cursor.active = false
        }
        node.active = true

        if (node.children) {
            node.toggled = toggled
        }
        this.setState({ cursor: node })

        if (node.children && node.children.length === 1) {
            this.onToggle(node.children[0], true)
        }
    }

    render() {
        return (
            <div>
                <t.Treebeard
                    data={ this.props.tree }
                    decorators={ t.decorators }
                    onToggle={ this.onToggle }
                    style={ globalStyle }
                />
            </div>
        )
    }
}
