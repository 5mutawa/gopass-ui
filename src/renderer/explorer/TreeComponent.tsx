import * as React from 'react'
import * as t from 'react-treebeard'
import * as m from 'react-materialize'
import { globalStyle } from './TreebeardStyle'
import credentialIconMappings, { CredentialIconMapping } from './CredentialIconMappings'

export interface Tree {
    name: string
    toggled?: boolean
    loading?: boolean
    children?: Tree[]
    entryId?: any
}

export interface TreeComponentProps {
    tree: Tree
    onLeafClick: (leafId: string) => void
}

const decorators = {
    ...t.decorators
}

decorators.Header = ({ style, node }: any) => {
    const iconStyle = { marginRight: '5px' }
    let iconType = 'folder'

    if (!node.children && node.entryId) {
        credentialIconMappings.forEach((mapping: CredentialIconMapping) => {
            const includesAnyCharacteristica = mapping.mustIncludeOnOf.find(needle =>
                node.name.includes(needle)
            )

            if (includesAnyCharacteristica) {
                iconType = mapping.icon
            }
        })

        if (iconType === 'folder') {
            iconType = 'comment'
        }
    }

    return (
        <div style={style.base}>
            <m.Icon small style={iconStyle}>
                {iconType}
            </m.Icon>

            {node.name}
        </div>
    )
}

export default class TreeComponent extends React.Component<TreeComponentProps, any> {
    constructor(props: TreeComponentProps) {
        super(props)
        this.state = {}
        this.onToggle = this.onToggle.bind(this)
    }

    onToggle(node: any, toggled: boolean) {
        if ((!node.children || node.children.length === 0) && !!node.entryId) {
            this.props.onLeafClick(node.entryId)
        }

        if (this.state.cursor) {
            this.state.cursor.active = false
        }
        node.active = true

        if (node.children) {
            node.toggled = toggled
        }
        this.setState({ cursor: node })
    }

    render() {
        return (
            <div>
                <t.Treebeard
                    data={this.props.tree}
                    decorators={decorators}
                    onToggle={this.onToggle}
                    style={globalStyle}
                />
            </div>
        )
    }
}
