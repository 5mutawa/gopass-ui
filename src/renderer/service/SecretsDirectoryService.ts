import { Tree } from '../components/explorer/TreeComponent'

export default class SecretsDirectoryService {
    public static secretPathsToTree(secretPaths: string[]): Tree {
        const directory = SecretsDirectoryService.secretPathsToDirectory(secretPaths)
        const tree = SecretsDirectoryService.directoryToTree(directory)
        return tree
    }

    /**
     * Convert
     *  from: "metro/service/ncr/cassandra/password"
     *  to: "{ metro: { service: { ncr: { cassandra: { password: {} } } } }  }"
     */
    private static secretPathsToDirectory(secretPaths: string[]): any {
        const directory: any = {}

        for (const secretPath of secretPaths) {
            const segments = secretPath.split('/')

            let tempDir = directory
            segments.forEach((segment: string, index: number) => {
                if (!tempDir[segment]) {
                    if (index + 1 === segments.length) {
                        tempDir[segment] = secretPath
                    } else {
                        tempDir[segment] = {}
                    }
                }

                tempDir = tempDir[segment]
            })
        }

        return directory
    }

    /**
     * Convert
     *  from: "{ metro: { service: { ncr: { cassandra: { password: {} } } } }  }"
     *  to Tree interface
     */
    private static directoryToTree(directory: any): Tree {
        const children: Tree[] = []
        const tree: Tree = {
            name: 'Stores',
            toggled: true,
            children: SecretsDirectoryService.getChildren(directory, true)
        }

        return tree
    }

    private static getChildren(
        directory: any | string,
        toggled: boolean = false
    ): Tree[] | undefined {
        if (!(directory instanceof Object)) {
            return undefined
        }
        const keys = Object.keys(directory).filter(key => key !== '')

        if (keys.length === 0) {
            return undefined
        }

        return keys.map(name => {
            const children = SecretsDirectoryService.getChildren(directory[name])
            return {
                name,
                children,
                toggled,
                entryId: children && children.length > 0 ? undefined : directory[name]
            }
        })
    }
}
