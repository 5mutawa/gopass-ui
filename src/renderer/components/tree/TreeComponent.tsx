import * as React from 'react'
import * as t from 'react-treebeard'
import { globalStyle } from './TreeStyle'
import { TreeHeader as Header } from './TreeHeader'

export interface Tree {
    name: string
    toggled?: boolean
    loading?: boolean
    children?: Tree[]
    entryId?: any
}

export interface TreeComponentProps {
    tree: Tree
    onLeafClick: (leafId: string) => void
}

export default class TreeComponent extends React.Component<TreeComponentProps, any> {
    public state: any = {}

    public render() {
        return <t.Treebeard
            data={this.props.tree}
            decorators={{ ...t.decorators, Header }}
            onToggle={this.onToggle}
            style={globalStyle}
        />
    }

    private onToggle = (node: any, toggled: boolean) => {
        if ((!node.children || node.children.length === 0) && !!node.entryId) {
            this.props.onLeafClick(node.entryId)
        }

        if (this.state.cursor) {
            this.state.cursor.active = false
        }
        node.active = true

        if (node.children) {
            node.toggled = toggled
        }
        this.setState({ cursor: node })

        if (node.children && node.children.length === 1) {
            this.onToggle(node.children[ 0 ], true)
        }
    }
}
