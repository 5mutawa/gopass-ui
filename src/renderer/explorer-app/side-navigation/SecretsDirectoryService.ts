import { Tree } from '../../components/tree/TreeComponent'

export default class SecretsDirectoryService {
    public static secretPathsToTree(secretPaths: string[], selectedSecretName: string | undefined): Tree {
        const directory = SecretsDirectoryService.secretPathsToDirectory(secretPaths)
        return SecretsDirectoryService.directoryToTree(directory, selectedSecretName, secretPaths.length)
    }

    /**
     * Convert
     *  from: "xyz/service/someServiceName/db/password"
     *  to: "{ xyz: { service: { someServiceName: { db: { password: {} } } } }  }"
     */
    private static secretPathsToDirectory(secretPaths: string[]): any {
        const directory: { [key: string]: any } = {}

        for (const secretPath of secretPaths) {
            const segments = secretPath.split('/')

            let tempDir = directory
            segments.forEach((segment: string, index: number) => {
                if (!tempDir[segment]) {
                    tempDir[segment] = index + 1 === segments.length ? secretPath : {}
                }

                tempDir = tempDir[segment]
            })
        }

        return directory
    }

    /**
     * Convert
     *  from: "{ xyz: { service: { someServiceName: { db: { password: {} } } } }  }"
     *  to Tree interface
     */
    private static directoryToTree(directory: any, selectedPath: string | undefined, totalEntries: number): Tree {
        const openAllEntries = totalEntries <= 15
        const children = SecretsDirectoryService.getChildren(directory, selectedPath, true, openAllEntries)
        const tree: Tree = {
            name: 'Stores',
            toggled: true,
            path: '',
            children
        }

        return tree
    }

    private static getChildren(
        directory: any | string,
        selectedPath: string | undefined,
        toggled: boolean = false,
        toggleAll: boolean = false,
        parentPath = ''
    ): Tree[] | undefined {
        if (!(directory instanceof Object)) {
            return undefined
        }
        const keys = Object.keys(directory).filter(key => key !== '')

        if (keys.length === 0) {
            return undefined
        }

        return keys.map(name => {
            const path = parentPath.length > 0 ? parentPath + '/' + name : name
            const children = SecretsDirectoryService.getChildren(directory[name], selectedPath, false, toggleAll, path)
            const isToggledBecauseChildSelected =
                !!selectedPath && selectedPath.length > 0 && !!path && path.length > 0 && !!children && children.length > 0 && selectedPath.startsWith(path)

            return {
                name,
                path,
                children,
                toggled: toggleAll || toggled || isToggledBecauseChildSelected
            }
        })
    }
}
