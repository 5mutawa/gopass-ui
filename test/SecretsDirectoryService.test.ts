import { Tree } from '../src/renderer/components/tree/TreeComponent'
import SecretsDirectoryService from '../src/renderer/explorer-app/side-navigation/SecretsDirectoryService'

describe('SecretsDirectoryService', () => {
    it('should transform a list of secret names into tree structure', () => {
        const secretPaths = [
            'codecentric/common/github/password',
            'codecentric/common/github/username',
            'codecentric/common/gitlab/password',
            'codecentric/common/gitlab/username',
            'codecentric/customers/some/notes'
        ]
        const tree: Tree = SecretsDirectoryService.secretPathsToTree(secretPaths, { name: 'someName', path: '' }, false)
        expect(tree).toMatchSnapshot()
    })

    it('should automatically toggle all nodes', () => {
        const secretPaths = [
            'codecentric/some-secret',
            'codecentric/common/something',
            'codecentric/common/another-thing',
            'codecentric/db/user',
            'codecentric/db/password'
        ]
        const tree: Tree = SecretsDirectoryService.secretPathsToTree(secretPaths, { name: 'someName', path: '' }, true)
        expect(tree).toMatchSnapshot()
    })

    it('should preserve previously toggled nodes when building a new tree', () => {
        const secretPaths = [
            'codecentric/some-secret',
            'codecentric/common/something',
            'codecentric/common/another-thing',
            'codecentric/db/user',
            'codecentric/db/password'
        ]
        const tree: Tree = SecretsDirectoryService.secretPathsToTree(
            secretPaths,
            {
                name: '',
                path: '',
                children: [
                    {
                        name: 'codecentric',
                        path: 'codecentric',
                        toggled: true,
                        children: [{ name: 'some-secret', path: 'codecentric/some-secret' }, { name: 'common', path: 'codecentric/common', toggled: true }]
                    }
                ]
            },
            false
        )
        expect(tree).toMatchSnapshot()
    })
})
