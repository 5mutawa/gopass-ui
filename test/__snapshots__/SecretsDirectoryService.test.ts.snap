// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecretsDirectoryService should automatically toggle all nodes 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": undefined,
          "name": "some-secret",
          "path": "codecentric/some-secret",
          "toggled": true,
        },
        Object {
          "children": Array [
            Object {
              "children": undefined,
              "name": "something",
              "path": "codecentric/common/something",
              "toggled": true,
            },
            Object {
              "children": undefined,
              "name": "another-thing",
              "path": "codecentric/common/another-thing",
              "toggled": true,
            },
          ],
          "name": "common",
          "path": "codecentric/common",
          "toggled": true,
        },
        Object {
          "children": Array [
            Object {
              "children": undefined,
              "name": "user",
              "path": "codecentric/db/user",
              "toggled": true,
            },
            Object {
              "children": undefined,
              "name": "password",
              "path": "codecentric/db/password",
              "toggled": true,
            },
          ],
          "name": "db",
          "path": "codecentric/db",
          "toggled": true,
        },
      ],
      "name": "codecentric",
      "path": "codecentric",
      "toggled": true,
    },
  ],
  "name": "Stores",
  "path": "",
  "toggled": true,
}
`;

exports[`SecretsDirectoryService should preserve previously toggled nodes when building a new tree 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": undefined,
          "name": "some-secret",
          "path": "codecentric/some-secret",
          "toggled": false,
        },
        Object {
          "children": Array [
            Object {
              "children": undefined,
              "name": "something",
              "path": "codecentric/common/something",
              "toggled": false,
            },
            Object {
              "children": undefined,
              "name": "another-thing",
              "path": "codecentric/common/another-thing",
              "toggled": false,
            },
          ],
          "name": "common",
          "path": "codecentric/common",
          "toggled": true,
        },
        Object {
          "children": Array [
            Object {
              "children": undefined,
              "name": "user",
              "path": "codecentric/db/user",
              "toggled": false,
            },
            Object {
              "children": undefined,
              "name": "password",
              "path": "codecentric/db/password",
              "toggled": false,
            },
          ],
          "name": "db",
          "path": "codecentric/db",
          "toggled": false,
        },
      ],
      "name": "codecentric",
      "path": "codecentric",
      "toggled": true,
    },
  ],
  "name": "Stores",
  "path": "",
  "toggled": true,
}
`;

exports[`SecretsDirectoryService should transform a list of secret names into tree structure 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": undefined,
                  "name": "password",
                  "path": "codecentric/common/github/password",
                  "toggled": false,
                },
                Object {
                  "children": undefined,
                  "name": "username",
                  "path": "codecentric/common/github/username",
                  "toggled": false,
                },
              ],
              "name": "github",
              "path": "codecentric/common/github",
              "toggled": false,
            },
            Object {
              "children": Array [
                Object {
                  "children": undefined,
                  "name": "password",
                  "path": "codecentric/common/gitlab/password",
                  "toggled": false,
                },
                Object {
                  "children": undefined,
                  "name": "username",
                  "path": "codecentric/common/gitlab/username",
                  "toggled": false,
                },
              ],
              "name": "gitlab",
              "path": "codecentric/common/gitlab",
              "toggled": false,
            },
          ],
          "name": "common",
          "path": "codecentric/common",
          "toggled": false,
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": undefined,
                  "name": "notes",
                  "path": "codecentric/customers/some/notes",
                  "toggled": false,
                },
              ],
              "name": "some",
              "path": "codecentric/customers/some",
              "toggled": false,
            },
          ],
          "name": "customers",
          "path": "codecentric/customers",
          "toggled": false,
        },
      ],
      "name": "codecentric",
      "path": "codecentric",
      "toggled": true,
    },
  ],
  "name": "Stores",
  "path": "",
  "toggled": true,
}
`;
